File: cough_android/app/src/main/java/com/twintipsolutions/cough/domain/recorder/AudioRecorder.kt
package com.twintipsolutions.cough.domain.recorder

import android.content.Context
import android.content.pm.PackageManager
import android.media.MediaRecorder
import android.os.Build
import android.util.Base64
import android.util.Log
import androidx.core.content.ContextCompat
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.functions.FirebaseFunctions
import com.twintipsolutions.cough.domain.model.CoughAnalysisResult
import com.twintipsolutions.cough.domain.model.PotentialCause
import com.twintipsolutions.cough.ui.components.AppReviewManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import java.io.File

class AudioRecorder(private val context: Context) {
    companion object {
        private const val TAG = "AudioRecorder"
        private const val MAX_RECORDING_DURATION = 10000L
    }

    private val functions = FirebaseFunctions.getInstance("us-central1")
    private val auth = FirebaseAuth.getInstance()

    private var mediaRecorder: MediaRecorder? = null
    private var audioFile: File? = null
    private var recordingStartTime: Long = 0

    private val _isRecording = MutableStateFlow(false)
    val isRecording: StateFlow<Boolean> = _isRecording.asStateFlow()

    private val _recordingTime = MutableStateFlow(0L)
    val recordingTime: StateFlow<Long> = _recordingTime.asStateFlow()

    private val _audioLevel = MutableStateFlow(0f)
    val audioLevel: StateFlow<Float> = _audioLevel.asStateFlow()

    private val _isAnalyzing = MutableStateFlow(false)
    val isAnalyzing: StateFlow<Boolean> = _isAnalyzing.asStateFlow()

    private val _analysisResult = MutableStateFlow<CoughAnalysisResult?>(null)
    val analysisResult: StateFlow<CoughAnalysisResult?> = _analysisResult.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    fun hasRecordingPermission(): Boolean {
        return ContextCompat.checkSelfPermission(
            context,
            android.Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED
    }

    suspend fun startRecording() {
        if (!hasRecordingPermission()) {
            _errorMessage.value = "Microphone permission is required to analyze your cough"
            return
        }

        withContext(Dispatchers.IO) {
            try {
                setupRecording()
                mediaRecorder?.start()
                recordingStartTime = System.currentTimeMillis()
                _isRecording.value = true
                _recordingTime.value = 0L
                _errorMessage.value = null

                startRecordingTimer()
                Log.d(TAG, "Recording started")
            } catch (e: Exception) {
                Log.e(TAG, "Failed to start recording", e)
                _errorMessage.value = "Failed to start recording: ${e.message}"
            }
        }
    }

    private suspend fun startRecordingTimer() {
        withContext(Dispatchers.IO) {
            while (_isRecording.value) {
                val elapsed = System.currentTimeMillis() - recordingStartTime
                _recordingTime.value = elapsed
                updateAudioLevel()

                if (elapsed >= MAX_RECORDING_DURATION) {
                    stopRecording()
                    break
                }
                delay(100)
            }
        }
    }

    private fun updateAudioLevel() {
        if (!_isRecording.value) {
            _audioLevel.value = 0f
        } else {
            _audioLevel.value = (Math.random() * 0.3f + 0.1f).toFloat()
        }
    }

    private fun setupRecording() {
        audioFile = File(context.cacheDir, "cough_${System.currentTimeMillis()}.m4a")

        mediaRecorder = if (Build.VERSION.SDK_INT >= 31) {
            MediaRecorder(context)
        } else {
            @Suppress("DEPRECATION")
            MediaRecorder()
        }.apply {
            setAudioSource(MediaRecorder.AudioSource.MIC)
            setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            setAudioSamplingRate(44100)
            setAudioChannels(1)
            setAudioEncodingBitRate(128000)
            setOutputFile(audioFile?.absolutePath)
            prepare()
        }
    }

    suspend fun stopRecording() {
        withContext(Dispatchers.IO) {
            try {
                mediaRecorder?.apply {
                    stop()
                    release()
                }
                mediaRecorder = null
                _isRecording.value = false
                _audioLevel.value = 0f

                Log.d(TAG, "Recording stopped. Duration: ${_recordingTime.value}ms")

                audioFile?.let { file ->
                    if (file.exists()) {
                        Log.d(TAG, "Audio file saved: ${file.absolutePath}, size: ${file.length()} bytes")
                        analyzeAudio(file)
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to stop recording", e)
                _errorMessage.value = "Failed to stop recording: ${e.message}"
            }
        }
    }

    private suspend fun analyzeAudio(file: File) {
        _isAnalyzing.value = true
        _analysisResult.value = null

        try {
            val audioBytes = file.readBytes()
            val base64Audio = Base64.encodeToString(audioBytes, Base64.NO_WRAP)

            val userId = auth.currentUser?.uid ?: ""
            val data = hashMapOf(
                "audio" to base64Audio,
                "userId" to userId,
                "platform" to "android",
                "appVersion" to getAppVersion()
            )

            val result = functions
                .getHttpsCallable("analyzeCough")
                .call(data)
                .continueWith { task ->
                    @Suppress("UNCHECKED_CAST")
                    task.result?.data as? Map<String, Any>
                }
                .addOnSuccessListener { response ->
                    if (response != null) {
                        try {
                            val analysisResult = parseAnalysisResult(response)
                            _analysisResult.value = analysisResult
                            // Trigger review prompt after successful analysis
                            AppReviewManager.incrementAnalysisCount(context)
                        } catch (e: Exception) {
                            Log.e(TAG, "Failed to parse analysis result", e)
                            _errorMessage.value = "Failed to parse analysis: ${e.message}"
                        }
                    }
                    _isAnalyzing.value = false
                }
                .addOnFailureListener { e ->
                    Log.e(TAG, "Analysis failed", e)
                    _errorMessage.value = "Analysis failed: ${e.message}"
                    _isAnalyzing.value = false
                }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to analyze audio", e)
            _errorMessage.value = "Failed to analyze: ${e.message}"
            _isAnalyzing.value = false
        }
    }

    private fun getAppVersion(): String {
        return try {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName ?: "1.0"
        } catch (e: Exception) {
            "1.0"
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseAnalysisResult(map: Map<String, Any>): CoughAnalysisResult {
        val results = map["results"] as? Map<String, Any>
            ?: throw Exception("Invalid response format: missing results")
        val insights = map["insights"] as? Map<String, Any>
            ?: throw Exception("Invalid response format: missing insights")

        val analysisId = (map["analysisId"] as? String) ?: ""
        val timestamp = (map["timestamp"] as? Double) ?: (System.currentTimeMillis() / 1000.0)

        val coughType = (results["coughType"] as? String) ?: "unknown"
        val severity = (results["severity"] as? String) ?: "mild"
        val characteristics = (results["characteristics"] as? List<*>)
            ?.filterIsInstance<String>() ?: emptyList()
        val potentialCauses = parsePotentialCauses(
            (results["potentialCauses"] as? List<*>) ?: emptyList()
        )
        val managementApproaches = (results["managementApproaches"] as? List<*>)
            ?.filterIsInstance<String>() ?: emptyList()
        val urgency = (results["urgency"] as? String) ?: "routine"
        val confidence = (results["confidence"] as? Double) ?: 0.5

        val soundPattern = (insights["soundPattern"] as? String) ?: ""
        val frequency = (insights["frequency"] as? String) ?: ""
        val duration = (insights["duration"] as? String) ?: ""
        val additionalNotes = (insights["additionalNotes"] as? List<*>)
            ?.filterIsInstance<String>() ?: emptyList()

        return CoughAnalysisResult(
            analysisId = analysisId,
            timestamp = timestamp,
            coughType = coughType,
            severity = severity,
            characteristics = characteristics,
            potentialCauses = potentialCauses,
            managementApproaches = managementApproaches,
            urgency = urgency,
            confidence = confidence,
            soundPattern = soundPattern,
            frequency = frequency,
            duration = duration,
            additionalNotes = additionalNotes
        )
    }

    @Suppress("UNCHECKED_CAST")
    private fun parsePotentialCauses(list: List<*>): List<PotentialCause> {
        return list.mapNotNull { item ->
            (item as? Map<String, Any>)?.let { map ->
                PotentialCause(
                    condition = (map["condition"] as? String) ?: "",
                    likelihood = (map["likelihood"] as? String) ?: "",
                    description = (map["description"] as? String) ?: ""
                )
            }
        }
    }

    fun cleanup() {
        mediaRecorder?.release()
        mediaRecorder = null
        audioFile?.delete()
    }
}

File: cough_ios/cough/AudioRecorder.swift
//
//  AudioRecorder.swift
//  cough
//
//  Created by Reid Glaze on 7/17/25.
//

import SwiftUI
import AVFoundation
import Firebase
import FirebaseAuth
import FirebaseFunctions

class AudioRecorder: NSObject, ObservableObject {
    @Published var isRecording = false
    @Published var recordingTime: TimeInterval = 0
    @Published var audioLevel: Float = 0
    @Published var isAnalyzing = false
    @Published var analysisResult: CoughAnalysisResult?
    @Published var errorMessage: String?
    
    private var audioRecorder: AVAudioRecorder?
    private var audioSession: AVAudioSession?
    private var recordingTimer: Timer?
    private var levelTimer: Timer?
    private let functions = Functions.functions(region: "us-central1")
    
    override init() {
        super.init()
        // Don't setup audio session immediately - wait until recording
    }
    
    private func setupAudioSession() {
        audioSession = AVAudioSession.sharedInstance()
        
        do {
            try audioSession?.setCategory(.playAndRecord, mode: .default)
            try audioSession?.setActive(true)
        } catch {
            print("Failed to set up audio session: \(error)")
            errorMessage = "Failed to set up audio recording"
        }
    }
    
    private func requestMicrophonePermission(completion: @escaping (Bool) -> Void) {
        // Request microphone permission
        if #available(iOS 17.0, *) {
            AVAudioApplication.requestRecordPermission { allowed in
                DispatchQueue.main.async {
                    if !allowed {
                        self.errorMessage = "Microphone access is required to analyze your cough"
                    }
                    completion(allowed)
                }
            }
        } else {
            audioSession?.requestRecordPermission { allowed in
                DispatchQueue.main.async {
                    if !allowed {
                        self.errorMessage = "Microphone access is required to analyze your cough"
                    }
                    completion(allowed)
                }
            }
        }
    }
    
    func startRecording() {
        // First setup audio session if not already done
        if audioSession == nil {
            setupAudioSession()
        }
        
        // Request permission first
        requestMicrophonePermission { [weak self] allowed in
            guard allowed else { return }
            
            self?.performRecording()
        }
    }
    
    private func performRecording() {
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let audioFilename = documentsPath.appendingPathComponent("cough_\(Date().timeIntervalSince1970).m4a")
        
        let settings = [
            AVFormatIDKey: Int(kAudioFormatMPEG4AAC),
            AVSampleRateKey: 44100,
            AVNumberOfChannelsKey: 1,
            AVEncoderAudioQualityKey: AVAudioQuality.high.rawValue
        ]
        
        do {
            audioRecorder = try AVAudioRecorder(url: audioFilename, settings: settings)
            audioRecorder?.delegate = self
            audioRecorder?.isMeteringEnabled = true
            audioRecorder?.record()
            
            isRecording = true
            recordingTime = 0
            
            // Start recording timer
            recordingTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
                self?.recordingTime += 0.1
                
                // Auto-stop after 10 seconds
                if self?.recordingTime ?? 0 >= 10 {
                    self?.stopRecording()
                }
            }
            
            // Start level monitoring
            levelTimer = Timer.scheduledTimer(withTimeInterval: 0.05, repeats: true) { [weak self] _ in
                self?.audioRecorder?.updateMeters()
                let level = self?.audioRecorder?.averagePower(forChannel: 0) ?? -160
                self?.audioLevel = max(0, (level + 160) / 160)
            }
        } catch {
            print("Failed to start recording: \(error)")
            errorMessage = "Failed to start recording"
        }
    }
    
    func stopRecording() {
        print("🛑 Stopping recording...")
        print("📊 Recording duration: \(recordingTime) seconds")
        audioRecorder?.stop()
        recordingTimer?.invalidate()
        levelTimer?.invalidate()
        isRecording = false
        audioLevel = 0
        
        // Process the recording
        if let url = audioRecorder?.url {
            print("📁 Recording saved at: \(url)")
            print("📁 File exists: \(FileManager.default.fileExists(atPath: url.path))")
            processRecording(at: url)
        } else {
            print("❌ No recording URL found")
        }
    }
    
    private func processRecording(at url: URL) {
        Task {
            await analyzeAudio(at: url)
        }
    }
    
    private func analyzeAudio(at url: URL) async {
        print("🔍 Starting audio analysis...")
        
        await MainActor.run {
            isAnalyzing = true
            errorMessage = nil
        }
        
        do {
            // Read audio file
            let audioData = try Data(contentsOf: url)
            print("📊 Audio file size: \(audioData.count) bytes")
            let base64Audio = audioData.base64EncodedString()
            
            // Get current user
            guard let userId = Auth.auth().currentUser?.uid else {
                throw NSError(domain: "AudioRecorder", code: 401, userInfo: [NSLocalizedDescriptionKey: "User not authenticated"])
            }
            print("👤 User ID: \(userId)")
            
            // Prepare request data
            let requestData: [String: Any] = [
                "userId": userId,
                "audioData": base64Audio,
                "audioFormat": "m4a",
                "duration": round(recordingTime * 10) / 10,  // Round to 1 decimal place
                "metadata": [
                    "recordedAt": Date().timeIntervalSince1970,
                    "platform": "iOS",
                    "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0"
                ]
            ]
            
            print("☁️ Calling Cloud Function: analyzeCough")
            print("📤 Request data size: \(requestData.description.count) characters")
            
            // Call Cloud Function
            let result = try await functions.httpsCallable("analyzeCough").call(requestData)
            
            print("✅ Cloud Function response received")
            print("📥 Response type: \(type(of: result.data))")
            
            // Parse result
            if let data = result.data as? [String: Any] {
                print("📋 Response data: \(data)")
                let analysisResult = try parseAnalysisResult(from: data)
            
                await MainActor.run {
                    self.analysisResult = analysisResult
                    self.isAnalyzing = false
                    print("🎉 Analysis complete!")
                }
                
                // Clean up audio file
                try? FileManager.default.removeItem(at: url)
            } else {
                print("❌ Invalid response format")
            }
        } catch {
            print("❌ Analysis error: \(error)")
            print("❌ Error type: \(type(of: error))")
            print("❌ Error details: \(error.localizedDescription)")
            
            await MainActor.run {
                self.errorMessage = "Analysis failed: \(error.localizedDescription)"
                self.isAnalyzing = false
            }
        }
    }
    
    private func parseAnalysisResult(from data: [String: Any]) throws -> CoughAnalysisResult {
        guard let results = data["results"] as? [String: Any],
              let insights = data["insights"] as? [String: Any] else {
            throw NSError(domain: "AudioRecorder", code: 500, userInfo: [NSLocalizedDescriptionKey: "Invalid response format"])
        }
        
        return CoughAnalysisResult(
            analysisId: data["analysisId"] as? String ?? "",
            timestamp: data["timestamp"] as? Double ?? Date().timeIntervalSince1970,
            coughType: results["coughType"] as? String ?? "unknown",
            severity: results["severity"] as? String ?? "mild",
            characteristics: results["characteristics"] as? [String] ?? [],
            potentialCauses: parsePotentialCauses(from: results["potentialCauses"] as? [[String: Any]] ?? []),
            managementApproaches: results["managementApproaches"] as? [String] ?? [],
            urgency: results["urgency"] as? String ?? "routine",
            confidence: results["confidence"] as? Double ?? 0.5,
            soundPattern: insights["soundPattern"] as? String ?? "",
            frequency: insights["frequency"] as? String ?? "",
            duration: insights["duration"] as? String ?? "",
            additionalNotes: insights["additionalNotes"] as? [String] ?? []
        )
    }
    
    private func parsePotentialCauses(from data: [[String: Any]]) -> [PotentialCause] {
        return data.compactMap { cause in
            guard let condition = cause["condition"] as? String,
                  let likelihood = cause["likelihood"] as? String,
                  let description = cause["description"] as? String else {
                return nil
            }
            return PotentialCause(condition: condition, likelihood: likelihood, description: description)
        }
    }
}

extension AudioRecorder: AVAudioRecorderDelegate {
    func audioRecorderDidFinishRecording(_ recorder: AVAudioRecorder, successfully flag: Bool) {
        if !flag {
            errorMessage = "Recording failed"
        }
    }
}

// Data Models
struct CoughAnalysisResult: Identifiable, Equatable {
    let id = UUID()
    let analysisId: String
    let timestamp: Double
    let coughType: String
    let severity: String
    let characteristics: [String]
    let potentialCauses: [PotentialCause]
    let managementApproaches: [String]
    let urgency: String
    let confidence: Double
    let soundPattern: String
    let frequency: String
    let duration: String
    let additionalNotes: [String]
    
    static func == (lhs: CoughAnalysisResult, rhs: CoughAnalysisResult) -> Bool {
        lhs.id == rhs.id
    }
    
    var urgencyColor: Color {
        switch urgency {
        case "urgent":
            return .red
        case "soon":
            return .orange
        default:
            return .green
        }
    }
    
    var severityColor: Color {
        switch severity {
        case "severe":
            return .red
        case "moderate":
            return .orange
        default:
            return .green
        }
    }
}

struct PotentialCause: Identifiable, Equatable {
    let id = UUID()
    let condition: String
    let likelihood: String
    let description: String
    
    var likelihoodColor: Color {
        switch likelihood {
        case "high":
            return .red
        case "medium":
            return .orange
        default:
            return .green
        }
    }
}